File: JAVA/1382. Balance a Binary Search Tree.java
import java.util.ArrayList;
import java.util.List;
class Solution {
    public TreeNode balanceBST(TreeNode root) {
        List<Integer> sortedElements = new ArrayList<>();
        collectInOrder(root, sortedElements);
        return constructBalancedBST(sortedElements, 0, sortedElements.size() - 1);
    }
    private void collectInOrder(TreeNode node, List<Integer> sortedElements) {
        if (node == null) {
            return;
        }
        collectInOrder(node.left, sortedElements);
        sortedElements.add(node.val);
        collectInOrder(node.right, sortedElements);
    }
    private TreeNode constructBalancedBST(List<Integer> elements, int start, int end) {
        if (start > end) {
            return null;
        }
        int mid = start + (end - start) / 2;
        TreeNode node = new TreeNode(elements.get(mid));
        node.left = constructBalancedBST(elements, start, mid - 1);
        node.right = constructBalancedBST(elements, mid + 1, end);
        return node;
    }
}
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: C++/1382. Balance a Binary Search Tree.cpp
class Solution {
    void inOrder(TreeNode* root,vector<int>& arr)
    {   
        if(!root)
            return;
        inOrder(root->left,arr);
        arr.push_back(root->val);
        inOrder(root->right,arr);
    }
    TreeNode* buildtree(vector<int>& sortedArr,int s,int e)
    {
        if(s>e)
            return NULL;
        int mid=s+(e-s)/2;
        int ele=sortedArr[mid];
        TreeNode* root=new TreeNode(ele);
        root->left=buildtree(sortedArr,s,mid-1);
        root->right=buildtree(sortedArr,mid+1,e);
        return root;
    }
public:
    TreeNode* balanceBST(TreeNode* root) {
        vector<int> sortedArr;
        inOrder(root,sortedArr);
        return buildtree(sortedArr,0,sortedArr.size()-1);
    }
};
